packages/cors/lib/cors.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * Adds AMP CORS headers.
+ *
+ * See also https://www.ampproject.org/docs/fundamentals/amp-cors-requests
+ *
+ * @param {Object} options
+ * @param {RegExp} options.sourceOriginPattern
+ * @return {Function} middleware function
+ */
+module.exports = (options) => {
+  options = options || {};
+  return (request, response, next) => {
+    // Get source origin from query
+    const sourceOrigin = request.query.__amp_source_origin;
+    if (!sourceOrigin) {
+      // it's not an AMP CORS request
+      next();
+      return;
+    }
+    // Check if sourceOrigin is allowed
+    if (options.sourceOriginPattern && !options.sourceOriginPattern.test(sourceOrigin)) {
+      response.status(403).end(); // forbidden
+      return;
+    }
+    // Get either the source origin or the cache origin
+    const origin = calculateOrigin_(request.headers, sourceOrigin);
+    if (!origin) {
```

Comment:
@andreban wondering if we should return `400` if __amp_source_origin, but neither `Origin` nor `AMP-SAME-ORIGIN`?

Reply:
> Not sure about how to best approach origin checking. Which caches should be whitelisted by default, cdn.ampproject.org/caches + bing?

Yes, I think that makes sense.

Reply:
> @andreban wondering if we should return `400` if __amp_source_origin, but neither `Origin` nor `AMP-SAME-ORIGIN`?

I think both are OK? You could see it as missing information that could generate a Bad Request, or information that doesn't give access to the content that would give Access Forbidden.

Reply:
returning a 400 now